### DISCUSSION TITLE
unable enter terminal when pudb.forked.set_trace

### DISCUSSION THREAD

**Discussion by Freed-Wu** [inducer/pudb#576] [Q&A]
```python
from multiprocessing import Process
def f(name):
    breakpoint()
    for i in range(3):
        print('hello', name)

p = Process(target=f, args=('bob',))
p.start()
p.join()
```

```sh
PYTHONBREAKPOINT=pudb.forked.set_trace python test.py
```
Then pudb stop in `breakpoint()`. Press `!` to enter terminal, it will

```
Warning: Input is not a terminal (fd=3).
Process Process-1:
Traceback (most recent call last):
  File "/usr/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/wzy/Desktop/test.py", line 4, in f
    for i in range(3):
  File "/home/wzy/Desktop/test.py", line 4, in f
    for i in range(3):
  File "/usr/lib/python3.10/bdb.py", line 90, in trace_dispatch
    return self.dispatch_line(frame)
  File "/usr/lib/python3.10/site-packages/pudb/debugger.py", line 221, in dispatch_line
    raise bdb.BdbQuit
bdb.BdbQuit
```

Is it a bug? Or I use wrong use method to debug multi process program? TIA!

**Comment by inducer** [comment]
I suspect the issue is that the `multiprocessing` workers do not have anything attached their stdin and stdout. `pudb.forked` (#402 by @jstriebel ) works/hacks around this for the situation while Urwid is active, but it has no control over the stdin/stdout that shells use. If someone is into low-level file descriptor hackery to make this go cleanly, I might consider it, but for now I think this is just a limitation. It might be best to stop attemtps to enter "external" consoles while in remote or forked debug.
